src/nfa/nfa-intersection.cc

```diff
@@ -119,27 +120,51 @@ class Intersection
     /**
      * Compute intersection preserving epsilon transitions.
      */
-    void compute_preserving_epsilon_transitions()
-    {
+    void compute_preserving_epsilon_transitions() {
         initialize_pairs_to_process();
 
-        if (lhs.trans_empty() || rhs.trans_empty()) { return; }
-
-        while (!pairs_to_process.empty())
-        {
+        while (!pairs_to_process.empty()) {
             pair_to_process = *pairs_to_process.begin();
             pairs_to_process.erase(pair_to_process);
+            compute_transitions_for_eps_pres_prod_for_state_pair();
+        }
+    }
 
-            // find all transitions that have same symbol for first and the second state in the pair_to_process
-            for (const auto& this_state_transitions: lhs.transitionrelation[pair_to_process.first] )
-            {
-                for (const auto& other_state_transitions: rhs.transitionrelation[pair_to_process.second])
-                {
-                    compute_for_state_pair_preserving_eps_trans(this_state_transitions,
-                                                                other_state_transitions);
+    /**
+     * Compute transitions for epsilon preserving product for a current state pair.
+     */
+    void compute_transitions_for_eps_pres_prod_for_state_pair() {
```

Comment:
The naming is weird. I didn't notice it before however the name is long and still doesn't provide a usefull information. Maybe it's problem that I don't have a domain knowledge needed to understand it but I would still like to discuss how to improve naming in intersection module.

Reply:
This function **computes and adds transitions** for the **current state pair** to the **epsilon preserving intersection**. Therefore, to shorten the name, `add_state_pair_trans_to_eps_pres_prod()` could be used. Do you find this suggestion better?

Reply:
what about `compute_transitions_for_statepair_eps_pres`?

Reply:
I would agree with `_` in `statepair` like so: `state_pair`.